File: swift/examples/formular/formular.swift
import Foundation
import Uimd

public final class FormApp: FormularUI
{
    private var wantsClose = false
    private(set) var action = ""

    override public func onSaveBtnClick()
    {
        action = "save"
        wantsClose = true
        requestClose()
    }

    override public func onCancelBtnClick()
    {
        action = "cancel"
        wantsClose = true
        requestClose()
    }

    override public func shouldClose() -> Bool
    {
        wantsClose
    }
}

@main
enum Program
{
    static func main()
    {
        let app = FormApp()
        let code = GeneratedWindowRuntime.runGeneratedAppMain
        {
            let result = GeneratedWindowRuntime.runGeneratedWindow(app, app.runtimeOptions(), CommandLine.arguments)
            if !app.action.isEmpty
            {
                printTerminalLines(formYaml(app, app.action))
            }
            return result
        }
        Foundation.exit(Int32(code))
    }

    private static func formYaml(_ window: FormularUI, _ action: String) -> String
    {
        var output = ""
        output += "action: \(action)\n"
        output += "form:\n"
        output += "  name: \(yamlScalar(window.name_input.value))\n"
        output += "  email: \(yamlScalar(window.email_input.value))\n"
        output += "  age: \(formatNumber(window.age_input.numberValue))\n"
        output += "  description: \(yamlScalar(window.description_input.value))\n"
        output += "  country: \(yamlScalar(window.country_combo.terminalText))\n"
        output += "  roles:\n"
        output += "  - \(yamlScalar(window.role_listbox.selectedValues.first ?? ""))\n"
        output += "  email_notifications: \(yamlBool(window.notify_check.checked))\n"
        output += "  accepted_terms: \(yamlBool(window.terms_check.checked))\n"
        return output
    }

    private static func yamlScalar(_ value: String) -> String
    {
        value.isEmpty ? "\"\"" : value
    }

    private static func yamlBool(_ value: Bool) -> String
    {
        value ? "true" : "false"
    }

    private static func formatNumber(_ value: Double) -> String
    {
        if value.rounded() == value
        {
            return String(Int(value))
        }
        return String(value)
    }

    private static func printTerminalLines(_ text: String)
    {
        FileHandle.standardOutput.write(Data("\r\n".utf8))
        for character in text
        {
            if character == "\n"
            {
                FileHandle.standardOutput.write(Data("\r\n".utf8))
            }
            else
            {
                FileHandle.standardOutput.write(Data(String(character).utf8))
            }
        }
    }
}

File: go/examples/formular/formular.go
package main

import (
	"fmt"
	"os"
	"strings"

	uimd "uimd"
)

type FormApp struct {
	*FormularUI
	action string
}

func NewFormApp() *FormApp {
	app := &FormApp{
		FormularUI: NewFormularUI(),
	}
	app.SetEventHandler(app)
	return app
}

func (app *FormApp) OnSaveBtnClick() {
	app.action = "save"
	app.RequestClose()
}

func (app *FormApp) OnCancelBtnClick() {
	app.action = "cancel"
	app.RequestClose()
}

func main() {
	code := uimd.RunGeneratedAppMain(func() int {
		app := NewFormApp()
		result := uimd.RunGeneratedWindow(app, app.RuntimeOptions(), os.Args)
		if app.action != "" {
			fmt.Print(formYaml(app.FormularUI, app.action))
		}
		return result
	})
	os.Exit(code)
}

func formYaml(window *FormularUI, action string) string {
	var builder strings.Builder
	builder.WriteString("action: ")
	builder.WriteString(action)
	builder.WriteString("\nform:\n")
	builder.WriteString("  name: ")
	builder.WriteString(yamlScalar(window.NameInput.Value))
	builder.WriteString("\n")
	builder.WriteString("  email: ")
	builder.WriteString(yamlScalar(window.EmailInput.Value))
	builder.WriteString("\n")
	builder.WriteString("  age: ")
	builder.WriteString(fmt.Sprintf("%g", window.AgeInput.NumberValue))
	builder.WriteString("\n")
	builder.WriteString("  description: ")
	builder.WriteString(yamlScalar(window.DescriptionInput.Value))
	builder.WriteString("\n")
	builder.WriteString("  country: ")
	builder.WriteString(yamlScalar(window.CountryCombo.Value))
	builder.WriteString("\n")
	builder.WriteString("  roles:\n")
	for _, role := range window.RoleListbox.SelectedValues() {
		builder.WriteString("  - ")
		builder.WriteString(yamlScalar(role))
		builder.WriteString("\n")
	}
	builder.WriteString("  email_notifications: ")
	builder.WriteString(yamlBool(window.NotifyCheck.Checked))
	builder.WriteString("\n")
	builder.WriteString("  accepted_terms: ")
	builder.WriteString(yamlBool(window.TermsCheck.Checked))
	builder.WriteString("\n")
	return builder.String()
}

func yamlScalar(value string) string {
	if value == "" {
		return "\"\""
	}
	return value
}

func yamlBool(value bool) string {
	if value {
		return "true"
	}
	return "false"
}
